### PR TITLE
Alternative approach to marking deferred expansion (for \the)

### PR DESCRIPTION
**Pull request by brucemiller**
Use a new catcode. Seems to me cleaner & more concise, as well as a little more intuitive what's happening when.

Hopefully it captures the same behaviour, and maybe even the correct behaviour. I've relied on the current set of tests, and haven't invented or added any new ones. It still fails texlive 2020, the same way.

This is a first layer; I'd like to try a couple of other experiments based on this.

### CODE REVIEW

**Comment by dginev** [comment]
Thanks for the alternative approach! My main hesitation about using a catcode remains on a principle debugging level - if a token marked after `\the`  "slips through" to a different argument passing context with this new catcode, for one reason or another (wrong parameter code, partial coverage/poor alignment with some yet-unstudied piece of TeX), it becomes, to me, a very major struggle to disentangle where that token got emitted, who set the flag, and why it wasn't handled. In expl3 that often entails reassembling dozens of intermediate macros by hand while doing prints.

It's also intuitive this _would_ happen, since TeX the program itself does not realize the behavior via a catcode, rather a global variable and a conditional to smuggle its contents through scan_toks. I'm almost tempted to suggest a second `{pushback_the}` in Gullet, as different direction for a refactor. Having to change `equals` in Token also looks concerning.

**Comment by brucemiller** [comment]
Yeah, crazy trade offs. The array approach makes `\the` errors explode in your face, which helps find & debug them, but otoh, makes debugging anything else going on in Gullet pretty impossible. And a new catcode is not Knuth's way, indeed, but neither are the arrays (other than the potential for leaks, the two approaches are very similar: weird tokens).  Somehow he manages to get away with "If it's `\the`, just don't expand it!", and everything works.  Somehow it seems as if it would only inhibit the *first* token.

File: lib/LaTeXML/Core/Gullet.pm [hunk 1]
```diff
@@ -396,12 +387,6 @@ sub readBalanced {
     Error('expected', "}", $self, "Gullet->readBalanced ran out of input in an unbalanced state.",
       $loc_message); }
   ## Performance enhancement:
```
Comment (dginev) [inline lib/LaTeXML/Core/Gullet.pm:389]: I got a chuckle from noticing that you left the "performance enhancement" part of the comment in, would be great if such comments enhanced performance on their own :>

**Comment by brucemiller** [comment]
As I said, I kept the parts I liked :>

**Comment by brucemiller** [comment]
Thanks for an interesting clue: Moving `PrepArgTokens` from Expansion to `DefPlain` and `DefExpanded` allows the later to remove the NOEXPAND flag! But requires all definers of macros (eg. `\newcommand`) to use `DefPlain instead of `{}`. I updated enough of them to get the expansion & expl3 tests to all pass (but *many* others still fail). And I'm not sure how much this localization really buys me. But interesting so far...

**Comment by dginev** [comment]
Yikes, committed here by accident. I still wanted the commit, but wanted it in a new branch as a PR comparison. One second, reverting...

**Comment by dginev** [comment]
Alright, damage undone, making a new branch PR...

**Comment by dginev** [comment]
> Thanks for an interesting clue: Moving `PrepArgTokens` from Expansion to `DefPlain` and `DefExpanded` allows the later to remove the NOEXPAND flag! But requires all definers of macros (eg. `\newcommand`) to use `DefPlain instead of `{}`. I updated enough of them to get the expansion & expl3 tests to all pass (but _many_ others still fail). And I'm not sure how much this localization really buys me. But interesting so far...
> 

You probably didn't axe enough infrastructure :> My PR gets all tests passing with very little use left of the new catcode. Avoiding it may be withing reach... 

> Addendum: or sneak a `prepped` option to DefMacro when you know that it's arguments have already been prepped, at least as a short-term testing fix.

Indeed, I sneaked a `noprep=>1` option to `Expandable::new`, which takes care of this.

**Pull request merged**